source/extensions/tracers/opencensus/opencensus_tracer_impl.cc

```diff
@@ -0,0 +1,268 @@
+#include "opencensus/trace/trace_config.h"
+#include "opencensus/trace/trace_params.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenCensus {
+
+namespace {
+
+class ConstantValues {
+public:
+  const Http::LowerCaseString TRACEPARENT{"traceparent"};
+  const Http::LowerCaseString GRPC_TRACE_BIN{"grpc-trace-bin"};
+  const Http::LowerCaseString X_CLOUD_TRACE_CONTEXT{"x-cloud-trace-context"};
+};
+
+typedef ConstSingleton<ConstantValues> Constants;
+
+/**
+ * OpenCensus tracing implementation of the Envoy Span object.
+ */
+class Span : public Tracing::Span {
+public:
+  Span(const Tracing::Config& config, const envoy::config::trace::v2::OpenCensusConfig& oc_config,
+       Http::HeaderMap& request_headers, const std::string& operation_name, SystemTime start_time,
+       const Tracing::Decision tracing_decision);
+
+  // Used by spawnChild().
+  Span(const envoy::config::trace::v2::OpenCensusConfig& oc_config,
+       ::opencensus::trace::Span&& span);
+
+  void setOperation(absl::string_view operation) override;
+  void setTag(absl::string_view name, absl::string_view value) override;
+  void log(SystemTime timestamp, const std::string& event) override;
+  void finishSpan() override;
+  void injectContext(Http::HeaderMap& request_headers) override;
+  Tracing::SpanPtr spawnChild(const Tracing::Config& config, const std::string& name,
+                              SystemTime start_time) override;
+  void setSampled(bool sampled) override;
+
+private:
+  ::opencensus::trace::Span span_;
+  const envoy::config::trace::v2::OpenCensusConfig& oc_config_;
+};
+
+::opencensus::trace::Span
+startSpanHelper(const std::string& name, bool traced, const Http::HeaderMap& request_headers,
+                const envoy::config::trace::v2::OpenCensusConfig& oc_config) {
+  // Determine if there is a parent context.
+  using OpenCensusConfig = envoy::config::trace::v2::OpenCensusConfig;
+  ::opencensus::trace::SpanContext parent_ctx;
+  for (const auto& incoming : oc_config.incoming_trace_context()) {
+    bool found = false;
+    switch (incoming) {
+    case OpenCensusConfig::trace_context: {
+      const Http::HeaderEntry* header = request_headers.get(Constants::get().TRACEPARENT);
+      if (header != nullptr) {
+        found = true;
+        parent_ctx = ::opencensus::trace::propagation::FromTraceParentHeader(
+            header->value().getStringView());
+      }
+      break;
+    }
+
+    case OpenCensusConfig::grpc_trace_bin: {
+      const Http::HeaderEntry* header = request_headers.get(Constants::get().GRPC_TRACE_BIN);
+      if (header != nullptr) {
+        found = true;
+        std::string val = std::string(header->value().getStringView());
+        // Add padding before decoding.
+        int pad = 4 - (val.size() % 4);
+        if (pad != 4) {
+          val += std::string(pad, '=');
+        }
+        parent_ctx = ::opencensus::trace::propagation::FromGrpcTraceBinHeader(Base64::decode(val));
```

Comment:
Can you add a variant of `Base64::decode` that doesn't require padding? From its internal structure it should not be very hard.

Reply:
What would you like the function prototype to look like / be called?